### PR TITLE
Errors on invalid feedback

### PR DESCRIPTION
**Pull request by cseelus**
Follow up on [issue 579](https://github.com/bootstrap-ruby/bootstrap_form/issues/579).

Took a little while, since I first wanted the project we use `bootstrap_form` with, to use Bootstrap 5 in a new development branch. So far the transition went smoother than the upgrade from 3 to 4, although it is far from complete.

The `bootstrap-5` branch of `bootstrap_form` works surprisingly well already.

### CODE REVIEW

**Review by lcreid — COMMENTED** [review]
This looks good so far. Could you please add some documentation in `UPGRADE-5.0.md` to describe this change and perhaps tell people what they would have to do if they can't or don't want to change their CSS from the previous version? Thanks in advance.

**Comment by cseelus** [comment]
I would add the following text to the bottom of `UPGRADE-5.0.md`:

---

## Different behavior for `errors_on` helper

The `errors_on` helper now wraps the error message in a CSS class `invalid-feedback`, instead of `alert` and `alert-danger`, as before.

This will display the error as any other [Bootstrap inline form error](https://getbootstrap.com/docs/5.0/forms/validation/#server-side), instead of displaying it as an [Bootstrap alert](https://getbootstrap.com/docs/5.0/components/alerts/).

---

Could you elaborate on what you mean by "tell people what they would have to do if they can't or don't want to change their CSS from the previous version"? Then I will add this also.
As far as I understand it, this should work without any extra CSS (as with < 5.0).

**Comment by lcreid** [comment]
> Could you elaborate on what you mean by "tell people what they would have to do if they can't or don't want to change their CSS from the previous version"? Then I will add this also.
> As far as I understand it, this should work without any extra CSS (as with < 5.0).

I'm thinking of the (unlikely) case where someone might have deliberately styled the CSS classes differently. Then they would have to do something to get the same results as before. And if they can't do anything about it, perhaps they should be able to. Maybe we could add a `class` option to `errors_on`, where they can specify a class or classes that override the default `invalid-feedback`. 

It may seem like an unlikely case, but I think it's good practice in general to give people a way to keep using our gem,. Another solution might be to document that if the new behaviour of `errors_on` is unacceptable, they can hard-code the HTML and CSS in place of using `errors_on`, or write their own method that replaces it. But I think adding a class argument to the method (and documenting it) feels like not too much effort. Let me know if I'm wrong.

What you wrote above is just perfect for what I originally asked for.

Thanks again for taking the time to contribute to `bootstrap_form`.

**Comment by cseelus** [comment]
So you mean an option like `class: 'foo'`, that will just override '.invalid-feedback'?

Regarding the custom helper idea you asked for, I played around with the idea for a bit:

```ruby
  def custom_errors_on(form, name, options = {})
    return if form.object.errors[name].blank?

    tag.div(class: 'alert alert-danger') do
      # hidden input tag is needed for the 'invalid-feedback' to be shown
      concat tag.input(type: 'hidden', class: 'is-invalid')
      concat form.errors_on(name, options)
    end
  end
```

**Comment by lcreid** [comment]
> So you mean an option like `class: 'foo'`, that will just override '.invalid-feedback'?

Yeah, basically something like that. BTW, I hope I'm not frustrating you, but I really don't have the cycles to think through the details, so I like to year what you propose. But, yeah, `class: 'foo'` is probably all we need here. (And see my comment at the end.)
> 
> Regarding the custom helper idea you asked for, I played around with the idea for a bit:
> 
> ```ruby
>   def custom_errors_on(form, name, options = {})
>     return if form.object.errors[name].blank?
> 
>     tag.div(class: 'alert alert-danger') do
>       # hidden input tag is needed for the 'invalid-feedback' to be shown
>       concat tag.input(type: 'hidden', class: 'is-invalid')
>       concat form.errors_on(name, options)
>     end
>   end
> ```

Thanks for doing the work to play with this. It seems to me like adding the `class` option to `errors_on`, plus documentation, isn't very much work compared to having to tell people how to write their own method to override ours. Maybe we could just add the `class` option? What do you think?

**Comment by cseelus** [comment]
I've added a custom class option for this helper.

File: lib/bootstrap_form/helpers/bootstrap.rb [hunk 1]
```diff
@@ -46,8 +46,9 @@ def errors_on(name, options={})
         return unless error?(name)
 
         hide_attribute_name = options[:hide_attribute_name] || false
+        custom_class_name   = options[:custom_class_name] || false
```
Comment (lcreid) [inline lib/bootstrap_form/helpers/bootstrap.rb:49]: I think the `_name` isn't really needed here, but that's not a strong opinion, so if you prefer to leave it, please do so.

File: UPGRADE-5.0.md [hunk 2]
```diff
@@ -11,3 +11,9 @@ Upgrading `bootstrap_form` to version 5 means you must upgrade your whole applic
 ## No `role="form"` Attribute
 
 As explained in #560, the `role="form"` attribute generated by `bootstrap_4` caused the W3C validator to output a warning. The `role="form"` attribute was deprecated in the 4.5.0 and is being remove completely in 5.0.0. This has no impact on `bootstrap_form` code itself, but may affect your application if it depended on a form having this attribute set. (Issue #569)
+
+## Different behavior for `errors_on` helper
+
+The `errors_on` helper now wraps the error message in a CSS class `invalid-feedback`, instead of `alert` and `alert-danger`, as before.
+
+This will display the error as any other [Bootstrap inline form error](https://getbootstrap.com/docs/5.0/forms/validation/#server-side), instead of displaying it as an [Bootstrap alert](https://getbootstrap.com/docs/5.0/components/alerts/).
```
Comment (lcreid) [inline UPGRADE-5.0.md:19]: What do you think about adding a short note here to tell people what to do if they need the old behaviour? In other words, tell them about the option they can now provide to `errors_on` to change the class name(s). 

If you add a description to the README, you could even just redirect them from here to the relevant documentation in the README.

**Review by lcreid — COMMENTED** [review]
Thanks. This looks good. 

Could you please add a simple test case to ensure the option generates the HTML you expect? Your change is simple and I'm sure it works, but we need the test to ensure that we don't break the behaviour in the future if, for example, we do some major refactoring or clean-up of the code.

We need to tell people about the new option in the README. There's a short section for `errors_on`. Just mention the option and perhaps give a quick example (maybe grab the test case).

Thanks again for taking the time to make this PR. We're almost there.

File: lib/bootstrap_form/helpers/bootstrap.rb (hunk 1 above)
Comment (cseelus) [reply lib/bootstrap_form/helpers/bootstrap.rb:49]: Yes, its better without `_name`

File: UPGRADE-5.0.md (hunk 2 above)
Comment (cseelus) [reply UPGRADE-5.0.md:19]: This should indeed be documented as well.

**Comment by cseelus** [comment]
Now that the approach seems to be ok, I added tests and more info about this change to the docs and into the changelog.

File: README.md [hunk 3]
```diff
@@ -803,7 +802,19 @@ You can hide the attribute name like this:
 Which outputs:
 
 ```html
-<div class="alert alert-danger">can't be blank.</div>
+<div class="invalid-feedback">can't be blank.</div>
```
Comment (lcreid) [inline README.md:805]: Nice. Clear and to the point. I like it.

File: UPGRADE-5.0.md [hunk 4]
```diff
@@ -11,3 +11,15 @@ Upgrading `bootstrap_form` to version 5 means you must upgrade your whole applic
...
+
+You can use the `custom_class` options for this helper with `alert alert-danger` to restore the old behaviour:
```
Comment (lcreid) [inline UPGRADE-5.0.md:21]: Again. Very clear.

File: test/bootstrap_form_test.rb [hunk 5]
```diff
@@ -793,8 +793,17 @@ class BootstrapFormTest < ActionView::TestCase
     @user.email = nil
     assert @user.invalid?
 
-    expected = '<div class="alert alert-danger">can&#39;t be blank, is too short (minimum is 5 characters)</div>'
+    expected = '<div class="invalid-feedback">can&#39;t be blank, is too short (minimum is 5 characters)</div>'
 
     assert_equivalent_xml expected, @builder.errors_on(:email, hide_attribute_name: true)
   end
+
+  test "errors_on use custom CSS classes" do
```
Comment (lcreid) [inline test/bootstrap_form_test.rb:801]: :+1:

**Review by lcreid — APPROVED** [review]
Looks great! I'll resolve the merge conflict and merge it. Thanks again for all your work on this.

**Pull request merged**